File: src/gestorAplicacion/enrutadorHFC/Router.java
package gestorAplicacion.enrutadorHFC;
import java.util.*;

import gestorAplicacion.host.Cliente;
import gestorAplicacion.servicio.Plano;

public class Router extends Cobertura {

  //ESTA CLASE ES UNA DE LAS PRINCIPALES DADO QUE ESTABLECE LA RELACIÓN ENTRE CLIENTE-PROVEEDOR Y ES UTILIZADA EN TODAS LAS FUNCIONALIDADES

  //ATRIBUTOS
  private final String IP;
  private int up;
  private int down;
  private boolean online;
  private int ping;
  private Plano coordenadas; 
  private Servidor servidorAsociado;
  private Antena antenaAsociada;
  private String sede;
  private int velocidad;

  //CONSTRUCTORES
  public Router(int up, int down, boolean online, Servidor servidorAsociado) {
    
    super(3);
    Random random = new Random();
    int octeto1 = random.nextInt(256);
    int octeto2 = random.nextInt(256);
    int octeto3 = random.nextInt(256);
    int octeto4 = random.nextInt(256);
    String ip = octeto1 + "." + octeto2 + "." + octeto3 + "." + octeto4;

    IP = ip;
    this.up = up;
    this.down = down;
    this.online = online;
    sede=servidorAsociado.getSede();   
    this.servidorAsociado=servidorAsociado;
    velocidad=250;
    servidorAsociado.getRouters().add(this);
  }

  public Router(int up, int down, boolean online, Antena antenaAsociada,Plano coordenadas,int g) {
    
    super(g);

    Random random = new Random();
    int octeto1 = random.nextInt(256);
    int octeto2 = random.nextInt(256);
    int octeto3 = random.nextInt(256);
    int octeto4 = random.nextInt(256);
    String ip = octeto1 + "." + octeto2 + "." + octeto3 + "." + octeto4;

    IP = ip;
    this.up = up;
    this.down = down;
    this.online = online;
    this.coordenadas=coordenadas;
    this.antenaAsociada=antenaAsociada;
    velocidad=250;
    sede=antenaAsociada.getSede();
  }

  public Router(){
    super(3);
    IP="";
    velocidad=250;
  }

  //METODOS

  //METODO DE INSTANCIA---FUNCIONALIDAD TEST-- SE ENCARGA DE ACTUALIZAR LAS MEGAS DEL CLIENTE DE ACUERDO CON LOS DISPOSITIVOS Y SI EL CLIENTE SE ENCUENTRA DENTRO O NO DE LA ZONA DE COBERTURA DE LA ANTENA QUE TIENE ASOCIADA
  public ArrayList<Integer> actualizarVelocidad(Cliente cliente) {
    int megasUp = cliente.getPlan().get(0);
    int megasDown = cliente.getPlan().get(1);
    Router routerCliente = cliente.getModem();
    ArrayList<Dispositivo> dispositivos = Dispositivo.getDispositivosTotales();
    ArrayList<Dispositivo> dispositivosCliente = verificarDispositivos(dispositivos, routerCliente);
    ArrayList<Integer> listaActualizada = new ArrayList<Integer>();
    if (dispositivosCliente.size() == 0) {
        megasUp=26;
        megasDown=18;
    } else {
      for (Dispositivo dispoActual : dispositivosCliente) {
       
          if (antenaAsociada.verificarZonaCobertura(this,antenaAsociada)){
            if (dispoActual.getNombre().equals("Celular")) {
              megasUp -= 4;
              megasDown -= 10;
            }
            if (dispoActual.getNombre().equals("Computador")) {
              megasUp -= 8;
              megasDown -= 20;
            }
            if (dispoActual.getNombre().equals("Televisor")) {
              megasUp -= 10;
              megasDown -= 30;
            }
          }else{
              megasUp=26;
              megasDown=18;
              break;
          }
        }

      }
      listaActualizada.add(megasDown);
      listaActualizada.add(megasUp);
      return listaActualizada;
    }

  //METODO INSTANCIA---FUNCIONALIDAD TEST---ACTUALIZA EL PING DE ACUERDO A LAS MEGAS
  public String actualizarPing(ArrayList<Integer> listaActualizada) {
    int up = listaActualizada.get(1);
    int down = listaActualizada.get(0);
    int ping = 0;
 
      if (up >= 60 && down >= 400) {
        ping = (int) (Math.random() * 20);
   
        return "\rPing: " + ping + " ms";

      } else if (up >= 40 && down >= 200) {
        ping = (int) (Math.random() * 21) + 20;
        return "\rPing: "+ ping+ " ms";
      } else if (up >= 20 && down >= 100) {
        ping = (int) (Math.random() * 21) + 41;
        return "\rPing: " + ping+" ms";
      } else if (up >= 0 && down >= 0) {
        ping = (int) (Math.random() * 21) + 90;
        return "\rPing: " + ping+" ms";
      } else if (up < 0 || down < 0) {
        ping = (int) (Math.random() * 99) + 900;
        return "\rPing: " + ping+" ms";
      }else{
        return null;
      }
  }

  //METODO DE ISNTANCIA--FUNCIONALIDAD TEST--SE ENCARGA DE LLAMAR OTROS METODOS Y RETORNAR UN RESULTADO DE ACUERDO A ESTO
  public String test(Cliente cliente) {
    ArrayList<Integer> up_down = cliente.getModem().actualizarVelocidad(cliente);

    int ping = Integer.parseInt(cliente.getModem().actualizarPing(up_down).split(" ")[1]);
   
    int up = up_down.get(1);
    int down = up_down.get(0);
    String resultado1 = " ";
    String resultado2 = " ";
    String resultado3 = " ";
    resultado1 = (900 < ping) ? "Red saturada": "Test de velocidad finalizado";
    resultado2 = (up_down.get(0)==18) ? "Motivo: Inaccesible a la zona de cobertura.":"Test de velocidad finalizado.";
    resultado3= (generacion==antenaAsociada.getGeneracion()) ? "Test de velocidad finalizado.":"Motivo: Generación incompatible";

    if (resultado1.contains("saturada")) {
      return actualizarPing(up_down) + "\n" + resultado1;
    } 
    else if(resultado2.contains("cobertura")){
      return resultado2;
    } 
    else if(resultado3.contains("incompatible")){
      return resultado3;
    }
    else{
      return "Ping: " + actualizarPing(up_down);
    }
  }

  //METODO INSTANCIA--FUNCIONALIDAD VISUALIZAR DISPOSITIVOS--SE ENCARGA DE VERIFICAR CUALES SON LOS DISPOSITIVOS QUE TIENE EL CLIENTE 
  public ArrayList<Dispositivo> verificarDispositivos(ArrayList<Dispositivo> dispositivosTotales, Router router) { // Crear
    ArrayList<Dispositivo> dispositivosConectados = new ArrayList<>();
    for (Dispositivo dispoActual : dispositivosTotales) {
      if (dispoActual.getIpAsociada().equals(router.getIP()) && (router.isOnline() == true)) {
        dispositivosConectados.add(dispoActual);
      }
    }
    return dispositivosConectados;
  }


  //METODO INSTANCIA--FUNCIONALIDAD MEJORA TU PLAN----SE ENCARGA DE CALCULAR DISTINTAS COSAS Y LLAMA OTRO METODO QUE HALLA EL SERVIDOR MÁS CERCANO Y CON MAYOR INTENSIDAD DE FLUJO
  public ArrayList<Object> protocoloSeleccionServidor(ArrayList<Servidor> servidores, Cliente c) {

    //1.COORDENADAS ROUTER CLIENTE
    int coordenadaX = coordenadas.getX();
    int coordenadaY = coordenadas.getY();

    //2.DISPOSITIVOS CONECTADOS AL ROUTER
    ArrayList<Dispositivo> dispositivosConectados = verificarDispositivos(Dispositivo.getDispositivosTotales(), this);

    //3.DISTANCIA DEL ROUTER AL SERVIDOR ASOCIADO
    int distancia = (int) Math.sqrt(Math.pow(servidorAsociado.getCoordenadas().getX() - coordenadaX, 2) + Math.pow(servidorAsociado.getCoordenadas().getY() - coordenadaY, 2));

    //4.VELOCIDAD TOTAL
    int velocidadTotal = new Conexion(IP, down, up, online, generacion, c,c.getModem().getServidorAsociado()).getVelocidad();

    //5.INTENSIDAD DE FLUJO INICIAL O ACTUAL
    intensidadFlujo = (int)((getServidorAsociado().getFLUJO_RED_NETO()+(velocidadTotal / dispositivosConectados.size()) - distancia)*getServidorAsociado().getPORCENTAJE_EFICIENCIA());

    //6.NOMBRE PLAN DEL CLIENTE
    String planCliente = c.getNombrePlan();

    //SE INVOCA EL METODO INTENSIDADFLUJOOPTIMA QUE ENCUENTRA EL SERVIDOR MÁS CERNCANO
    ArrayList<Object> intensidadMayor = intensidadFlujoOptima(servidores, c);
    Servidor servidorAproximado = (Servidor)(intensidadMayor.get(0)); 
    int intensidadServidorAproximado = (int)(intensidadMayor.get(1));

    //CARACTERISTICAS DE LA RECOMENDACION PARA RETORNAR
    ArrayList<Object> caracteristicas = new ArrayList<>();

    //SE REALIZAN LAS COMPARACIONES RESPECTIVAS
     if ( intensidadServidorAproximado == intensidadFlujo) {
      if (servidorAproximado == servidorAsociado) { //CUANDO LAS INTENSIDADES SON IGUALES Y SE TRATA DEL MISMO SERVIDOR (EL CLIENTE YA TIENE LA MEJOR OPCION)
        return null;
      } else { //CUANDO LAS INTENSIDADES COINCIDEN PERO EL SERVIDOR ES DISTINTO---SE COMPARA EL PRECIO DEL PLAN DE ACUERDO AL TIPO DEL PLAN DEL CLIENTE
        if (planCliente.equals("BASIC")){
          if(c.getPlan().get(2)>servidorAproximado.getProveedor().getPlanes().getBASIC().get(2)){

            caracteristicas.add(intensidadFlujo);
            caracteristicas.add(servidorAproximado);
            caracteristicas.add(servidorAproximado.getProveedor());
            caracteristicas.add(intensidadServidorAproximado);
            caracteristicas.add(servidorAproximado.getProveedor().getPlanes().getBASIC());
            
            return caracteristicas;
          }
    
        } else if (planCliente.equals("STANDARD")) {
          
          if(c.getPlan().get(2)>servidorAproximado.getProveedor().getPlanes().getSTANDARD().get(2)){

            caracteristicas.add(intensidadFlujo);
            caracteristicas.add(servidorAproximado);
            caracteristicas.add(servidorAproximado.getProveedor());
            caracteristicas.add(intensidadServidorAproximado);
            caracteristicas.add(servidorAproximado.getProveedor().getPlanes().getSTANDARD());
            
            return caracteristicas;
          }
  
        } else if (planCliente.equals("PREMIUM")) {
          
          if(c.getPlan().get(2)>servidorAproximado.getProveedor().getPlanes().getPREMIUM().get(2)){

            caracteristicas.add(intensidadFlujo);
            caracteristicas.add(servidorAproximado);
            caracteristicas.add(servidorAproximado.getProveedor());
            caracteristicas.add(intensidadServidorAproximado);
            caracteristicas.add(servidorAproximado.getProveedor().getPlanes().getPREMIUM());
            
            return caracteristicas;
          }
        }
        
      } 

    } else if (intensidadFlujo < intensidadServidorAproximado) { //CUANDO LA INTENSIDAD ACTUAL ES MENOR A LA CALCULADA CON EL SERVIDOR MÁS CERCANO
      if (planCliente.equals("BASIC")) {
        
        caracteristicas.add(intensidadFlujo);
        caracteristicas.add(servidorAproximado);
        caracteristicas.add(servidorAproximado.getProveedor());
        caracteristicas.add(intensidadServidorAproximado);
        caracteristicas.add(servidorAproximado.getProveedor().getPlanes().getBASIC());
        
        return caracteristicas;

      } else if (planCliente.equals("STANDARD")) {
        
        caracteristicas.add(intensidadFlujo);
        caracteristicas.add(servidorAproximado);
        caracteristicas.add(servidorAproximado.getProveedor());
        caracteristicas.add(intensidadServidorAproximado);
        caracteristicas.add(servidorAproximado.getProveedor().getPlanes().getSTANDARD());
        
        return caracteristicas;

      } else if (planCliente.equals("PREMIUM")) {
        
        caracteristicas.add(intensidadFlujo);
        caracteristicas.add(servidorAproximado);
        caracteristicas.add(servidorAproximado.getProveedor());
        caracteristicas.add(intensidadServidorAproximado);
        caracteristicas.add(servidorAproximado.getProveedor().getPlanes().getPREMIUM());
        
        return caracteristicas;
      }
    }
    return null;
  }

  //METODO DE INSTANCIA-- IMPLEMENTACION DEL METODO ABSTRACTO HEREDADO DE COBERTURA---BUSCA EL SERVIDOR MÁS CERCANO Y POR TANTO CON MAYOR INTENSIDAD DE FLUJO--FUNCIONALIDAD MEJORA TU PLAN
  public ArrayList<Object> intensidadFlujoOptima(ArrayList<Servidor> servidoresLocalidad, Cliente c) {

    //1.DISTANCIA DEL ROUTER AL SERVIDOR ASOCIADO INICIALMENTE
    int distancia = (int) Math.sqrt(Math.pow(servidorAsociado.getCoordenadas().getX() - coordenadas.getX(), 2) + Math.pow(servidorAsociado.getCoordenadas().getY() - coordenadas.getY(), 2));
    
    //APLICACION DE LIGADURA DINÁMICA Y SE HALLA LA VELOCIDAD TOTAL DE CONEXION
    Router conexion = new Conexion(IP, down, up, online, generacion, c, c.getModem().getServidorAsociado());
    int velocidadTotal = conexion.getVelocidad();
    
    //VARIABLES
    int distanciaTemporal = 0;
    Servidor servTemporal = null;

    //SE RECORREN LOS SERVIDORES DE LA SEDE Y SE HALLA EL MÁS CERCANO
    for (Servidor servidor : servidoresLocalidad) {

      distanciaTemporal = (int) Math.sqrt(Math.pow(servidor.getCoordenadas().getX() - coordenadas.getX(), 2) + Math.pow(servidor.getCoordenadas().getY() - coordenadas.getY(), 2));

      if (distanciaTemporal <= distancia) {
        distancia = distanciaTemporal;
        servTemporal = servidor;
      }
    }

    Integer intensidadFlujoMayor = (int)((servTemporal.getFLUJO_RED_NETO()+(velocidadTotal/verificarDispositivos(Dispositivo.getDispositivosTotales(), this).size()) - distancia)* servTemporal.getPORCENTAJE_EFICIENCIA());
    ArrayList<Object> lista = new ArrayList<Object>();
    
    lista.add(servTemporal);
    lista.add(intensidadFlujoMayor);
    
    return lista;
  }


  //METODOS INSTANCIA-- IMPLEMENTACION DEL METODO ABSTRACTO HEREDADO DE COBERTURA--FUNCIONALIDAD REPORTE
  public ArrayList<Integer> intensidadFlujoClientes(ArrayList<Cliente> ctes, Servidor servidor, boolean Reales) { //SI REALES==TRUE CALCULA LAS INTENSIDADES REALES, DE LO CONTRARIO LAS QUE DEBERIAN LLEGAR
   
    ArrayList<Integer> lista = new ArrayList<>();

    for (Cliente c : ctes) {

      int intensidad=0;
      String ip = c.getModem().getIP();
      int up = c.getModem().getUp();
      int down = c.getModem().getDown();
      boolean online = c.getModem().isOnline();
      int generacion = c.getModem().getGeneracion();
      int velocidad = new Conexion(ip, up, down, online, generacion, c, servidor).getVelocidad();
      int dispositivos = c.getModem().verificarDispositivos(Dispositivo.getDispositivosTotales(), c.getModem()).size();
      int distancia = (int) Math.sqrt(Math.pow(servidor.getCoordenadas().getX() - c.getModem().getCoordenadas().getX(), 2) + Math.pow(servidor.getCoordenadas().getY() - c.getModem().getCoordenadas().getY(), 2));
      double porcentaje = servidor.getPORCENTAJE_EFICIENCIA();

      if(dispositivos!=0){
        intensidad = (int)((servidor.getFLUJO_RED_NETO()+velocidad / dispositivos - distancia) * porcentaje);
      }else{
        intensidad = (int)((servidor.getFLUJO_RED_NETO()- distancia) * porcentaje);
      }
      
      int intensidad1 = (int)(intensidad/porcentaje);
      
      if (Reales) {
        lista.add(intensidad); //INTENSIDAD REAL (LAS QUE ESTÁ LLEGANDO)
      } else {
        lista.add(intensidad1); //INTENSIDAD ADECUADA (LA QUE DEBERIA LLEGAR)
      }
    }
    return lista;
  }

  //DEFINICIÓN DEL METODO ABSTRACTO QUE HEREDA DE COBERTURA--EVITAR ERRORES DE COMPILACION
  public Antena rastrearGeneracionCompatible(ArrayList<Antena> antenasSede,Router r){
    return null;
  }

  //METODO ESTATICO QUE CREA UN ROUTER PARA EL ADMINISTRADOR---UN ROUTER ALEATORIO--FUNCIONALIDAD REPORTE
  public static Router adminRouter(){
    return new Router();
  }

  //SETTERS Y GETTERS

  public String getIP() {
    return IP;
  }

  public int getUp() {
    return up;
  }

  public void setUp(int up) {
    this.up = up;
  }

  public int getDown() {
    return down;
  }

  public void setDown(int down) {
    this.down = down;
  }

  public boolean isOnline() {
    return online;
  }

  public void setOnline(boolean online) {
    this.online = online;
  }

  public int getPing() {
    return ping;
  }

  public void setPing(int ping) {
    this.ping = ping;
  }

  public Plano getCoordenadas() {
    return coordenadas;
  }

  public void setCoordenadas(Plano coordenadas) {
    this.coordenadas = coordenadas;
  }

  public Servidor getServidorAsociado() {
    return servidorAsociado;
  }

  public void setServidorAsociado(Servidor servidorAsociado) {
    this.servidorAsociado = servidorAsociado;
  }

  public Antena getAntenaAsociada() {
    return antenaAsociada;
  }

  public void setAntenaAsociada(Antena antenaAsociada) {
    this.antenaAsociada = antenaAsociada;
  }

  public String getSede() {
    return sede;
  }

  public void setSede(String sede) {
    this.sede = sede;
  }

  public int getVelocidad() {
    return velocidad;
  }

  public void setVelocidad(int velocidad) {
    this.velocidad = velocidad;
  }
  
}

File: Python/src/gestorAplicacion/enrutadorHFC/router.py
from random import randint
from math import sqrt,pow 
from dispositivo import Dispositivo
from cobertura import Cobertura

class Router(Cobertura):

    def __init__(self,up=0,down=0,online=False,servidor=None,antena=None,coordenadas=None,generacion=3):
        
        super().__init__(generacion)
        
        octeto1 = randint(100,256)
        octeto2 = randint(100,256)
        octeto3 = randint(100,256)
        octeto4 = randint(100,256)
        
        self._IP = str(octeto1) + "." + str(octeto2) + "." + str(octeto3) + "." + str(octeto4)

        self._up=up
        self._down=down
        self._online=online
        self._ping=0 
        self._coordenadas=coordenadas
        self._servidorAsociado=servidor
        self._antenaAsociada=antena
        self._velocidad=250

        if self._antenaAsociada!=None:
            self._sede=self._antenaAsociada.getSede()
            
        if self._servidorAsociado!=None: 
            self._servidorAsociado.getRouters().append(self)

    
    def actualizarVelocidad(self,cliente):

        megasUp = cliente.getPlan()[0]
        megasDown = cliente.getPlan()[1]
        routerCliente = cliente.getModem()
        dispositivos = Dispositivo.getDispositivosTotales()
        dispositivosCliente = self.verificarDispositivos(dispositivos,routerCliente)
        listaActualizada = []

        if(len(dispositivosCliente)==0):
            megasUp=26
            megasDown=18
        else:

            for dispoActual in dispositivosCliente:
                if(self._antenaAsociada.verificarZonaCobertura(self,self._antenaAsociada)):
                    if(dispoActual.getNombre() == "Celular"):
                        megasUp -= 4
                        megasDown -= 10
                    elif(dispoActual.getNombre() == "Computador"):
                        megasUp -= 8
                        megasDown -= 20
                    elif(dispoActual.getNombre() == "Televisor"):
                        megasUp -= 10
                        megasDown -= 30
                else:
                    megasDown=18
                    megasUp=26
                    break


        listaActualizada.append(megasDown)
        listaActualizada.append(megasUp)
        return listaActualizada

    
    def actualizarPing(self, listaActualizada):

        up = listaActualizada[1]
        down = listaActualizada[0]
        global ping 

        if(up >= 60 and down >= 400):
            ping = randint(0,20)
            return "Ping: " + str(ping) + " ms"

        elif(up >= 40 and down >= 200):
            ping = randint(21,40) 
            return "Ping: " + str(ping) + " ms"
        
        elif(up >= 20 and down >= 100):
            ping = randint(41,60)
            return "Ping: " + str(ping) + " ms"

        elif(up >= 0 and down >= 0):
            ping = randint(90,110)
            return "Ping: " + str(ping) + " ms"
        
        elif(up < 0 or down < 0):
            ping = randint(900,999)
            return "Ping: " + str(ping) + " ms" 
        
        else:
            return None

    
    def test(self, cliente):
        
        up_down = cliente.getModem().actualizarVelocidad(cliente)
        ping=int(cliente.getModem().actualizarPing(up_down).split(" ")[1])
        up = up_down[1]
        down = up_down[0]
        resultado1 = ""
        resultado2 = ""
        resultado3 = ""

        resultado1 = "Red saturada" if (900 < ping) else "Test de velocidad finalizado"
        resultado2 = "Inaccesible a la zona de cobertura" if up_down[0]==18 else "Test de velocidad finalizado"
        resultado3 = "Test de velocidad finalizado" if self.generacion==self._antenaAsociada.getGeneracion() else "Generacion incompatible"


        if("cobertura" in resultado2):
            return resultado2

        elif("saturada" in resultado1):
            return self.actualizarPing(up_down) + "\n" + resultado1

        elif("incompatible" in resultado3):
            return resultado3

        else:
            ping = self.actualizarPing(up_down)
            return ping

    
    def verificarDispositivos(self, dispositivosTotales,router):

        dispositivosConectados = []

        for dispoActual in dispositivosTotales :
            if(dispoActual.getIpAsociada() == router.getIP() and router.isOnline() == True ):
                dispositivosConectados.append(dispoActual)

        return dispositivosConectados

    

    def protocoloSeleccionServidor(self, servidores,c):

        from conexion import Conexion

        #1.COORDENADAS ROUTER CLIENTE
        coordenadaX = self._coordenadas.getX()
        coordenadaY = self._coordenadas.getY()

        #2.DISPOSITIVOS CONECTADOS AL ROUTER
        dispositivosConectados = self.verificarDispositivos(Dispositivo.getDispositivosTotales(),self)

        #3.DISTANCIA DEL ROUTER AL SERVIDOR ASOCIADO
        distancia = int(sqrt(pow(self._servidorAsociado.getCoordenadas().getX() - coordenadaX, 2) + pow(self._servidorAsociado.getCoordenadas().getY() - coordenadaY, 2)))

        #4.VELOCIDAD TOTAL
        velocidadTotal = Conexion(self._IP,self._down,self._up,self._online,self.generacion,c,self._servidorAsociado).getVelocidad()

        #5.INTENSIDAD DE FLUJO INICIAL O ACTUAL
        intensidadFlujo = int((self._servidorAsociado.getFLUJO_RED_NETO() + (velocidadTotal/len(dispositivosConectados)) - distancia) * self._servidorAsociado.getPORCENTAJE_EFICIENCIA())

        #6.NOMBRE PLAN DEL CLIENTE
        planCliente = c.getNombrePlan()

        intensidadMayor = self.intensidadFlujoOptima(servidores, c)

        servidorAproximado = intensidadMayor[0]
        intensidadServidorAproximado = int(intensidadMayor[1])

        #CARACTERISTICAS DE LA RECOMENDACION PARA RETORNAR
        caracteristicas = []

        if(intensidadServidorAproximado == intensidadFlujo):
        
            if(servidorAproximado == self._servidorAsociado):
                return None
            else:

                if(planCliente == "BASIC"):
                    if(c.getPlan()[2] > servidorAproximado.getProveedor().getPlanes().getBASIC()[2]):
                        
                        caracteristicas.append(intensidadFlujo)
                        caracteristicas.append(servidorAproximado)
                        caracteristicas.append(servidorAproximado.getProveedor())
                        caracteristicas.append(intensidadServidorAproximado)
                        caracteristicas.append(servidorAproximado.getProveedor().getPlanes().getBASIC())
                        return caracteristicas
                    
                elif(planCliente == "STANDARD"):
                    if(c.getPlan()[2] > servidorAproximado.getProveedor().getPlanes().getSTANDARD()[2]):
                        
                        caracteristicas.append(intensidadFlujo)
                        caracteristicas.append(servidorAproximado)
                        caracteristicas.append(servidorAproximado.getProveedor())
                        caracteristicas.append(intensidadServidorAproximado)
                        caracteristicas.append(servidorAproximado.getProveedor().getPlanes().getSTANDARD())
                        return caracteristicas
                    
                elif(planCliente == "PREMIUM"):
                    if(c.getPlan()[2] > servidorAproximado.getProveedor().getPlanes().getPREMIUM()[2]):
                        
                        caracteristicas.append(intensidadFlujo)
                        caracteristicas.append(servidorAproximado)
                        caracteristicas.append(servidorAproximado.getProveedor())
                        caracteristicas.append(intensidadServidorAproximado)
                        caracteristicas.append(servidorAproximado.getProveedor().getPlanes().getPREMIUM())
                        return caracteristicas
            
        elif(intensidadFlujo < intensidadServidorAproximado):
            if(planCliente == "BASIC"):

                caracteristicas.append(intensidadFlujo)
                caracteristicas.append(servidorAproximado)
                caracteristicas.append(servidorAproximado.getProveedor())
                caracteristicas.append(intensidadServidorAproximado)
                caracteristicas.append(servidorAproximado.getProveedor().getPlanes().getBASIC())
                return caracteristicas
            
            elif(planCliente == "STANDARD"):

                caracteristicas.append(intensidadFlujo)
                caracteristicas.append(servidorAproximado)
                caracteristicas.append(servidorAproximado.getProveedor())
                caracteristicas.append(intensidadServidorAproximado)
                caracteristicas.append(servidorAproximado.getProveedor().getPlanes().getSTANDARD())
                return caracteristicas
            
            elif(planCliente == "PREMIUM"):

                caracteristicas.append(intensidadFlujo)
                caracteristicas.append(servidorAproximado)
                caracteristicas.append(servidorAproximado.getProveedor())
                caracteristicas.append(intensidadServidorAproximado)
                caracteristicas.append(servidorAproximado.getProveedor().getPlanes().getPREMIUM())
                return caracteristicas
            
        return None
    

    def intensidadFlujoOptima(self, servidoresLocalidad, c):

        from conexion import Conexion

        distancia = int(sqrt(pow(self._servidorAsociado.getCoordenadas().getX() - self._coordenadas.getX(), 2) + pow(self._servidorAsociado.getCoordenadas().getY() - self._coordenadas.getY(), 2)))

        conexion = Conexion(self._IP, self._down, self._up, self._online, self.generacion, c, c.getModem().getServidorAsociado())
        velocidadTotal = conexion.getVelocidad()

        distanciaTemporal = 0
        servidorTemporal = None

        for servidor in servidoresLocalidad:
        
            distanciaTemporal = int (sqrt(pow(servidor.getCoordenadas().getX() - self._coordenadas.getX(), 2) + pow(servidor.getCoordenadas().getY() - self._coordenadas.getY(), 2)))
            
            if(distanciaTemporal <= distancia):
                distancia = distanciaTemporal
                servidorTemporal = servidor
        
        intensidadFlujoMayor = int((servidorTemporal.getFLUJO_RED_NETO()+(velocidadTotal/len(self.verificarDispositivos(Dispositivo.getDispositivosTotales(), self))) - distancia)* servidorTemporal.getPORCENTAJE_EFICIENCIA())
        return [servidorTemporal,intensidadFlujoMayor]     


    def intensidadFlujoClientes(self, ctes, servidor,reales):

        from conexion import Conexion
        
        lista = []

        for c in ctes:
            intensidad=0
            ip = c.getModem().getIP()
            up = c.getModem().getUp()
            down = c.getModem().getDown()
            online = c.getModem().isOnline()
            generacion = c.getModem().getGeneracion()
            velocidad = Conexion(ip, up, down, online, generacion, c, servidor).getVelocidad()
            dispositivos = len(c.getModem().verificarDispositivos(Dispositivo.getDispositivosTotales(), c.getModem()))
            distancia = int(sqrt(pow(servidor.getCoordenadas().getX() - c.getModem().getCoordenadas()[0], 2) + pow(servidor.getCoordenadas().getY() - c.getModem().getCoordenadas()[1], 2)))
            porcentaje = servidor.getPORCENTAJE_EFICIENCIA()

            if(dispositivos!=0):
                intensidad = int((servidor.getFLUJO_RED_NETO()+velocidad / dispositivos - distancia) * porcentaje)
            else:
                intensidad = int((servidor.getFLUJO_RED_NETO()- distancia) * porcentaje)
            
            intensidad1 = int(intensidad/porcentaje)

            if reales:
                lista.append(intensidad)
            else:
                lista.append(intensidad1)

        return lista         


    def rastrearGeneracionCompatible(self, antenasSede,r):
        return None
    
    @classmethod
    def adminRouter(cls):
        return Router()
    
    #SETTERS Y GETTERS

    def getIP(self):
        return self._IP
    
    def getUp(self):
        return self._up

    def setUp(self, up):
        self._up = up

    def getDown(self):
        return self._down
    
    def setDown(self, down):
        self._down = down
    
    def isOnline(self):
        return self._online

    def setOnline(self, online):
        self._online = online
    
    def getPing(self):
        return self._ping

    def setPing(self, ping):
        self._ping = ping
    
    def getCoordenadas(self):
        return self._coordenadas.crearPunto()
    
    def setCoordenadas(self, coordenadas):
        self._coordenadas = coordenadas
    
    def getServidorAsociado(self, ):
        return self._servidorAsociado

    def setServidorAsociado(self, servidorAsociado):
        self._servidorAsociado = servidorAsociado
    
    def getAntenaAsociada(self):
        return self._antenaAsociada
    
    def setAntenaAsociada(self, antenaAsociada):
        self._antenaAsociada = antenaAsociada

    def getSede(self):
        return self._sede
    
    def setSede(self, sede):
        self._sede = sede
    
    def getVelocidad(self):
        return self._velocidad
    
    def setVelocidad(self, velocidad):
        self._velocidad = velocidad